File: acoustic_iso_lib/seis_utils/seis_utils_float/src/dataCompMain.cpp
#include <omp.h>
#include <iostream>
#include <time.h>
#include <stdlib.h>
#include <stdio.h>
#include "float1DReg.h"
#include "float2DReg.h"
#include "float3DReg.h"
#include "float4DReg.h"
#include "ioModes.h"
#include <vector>
#include <string>

using namespace SEP;

int main(int argc, char **argv) {


    ioModes modes(argc, argv);
	std::shared_ptr <SEP::genericIO> io = modes.getDefaultIO();
	std::shared_ptr <paramObj> par = io->getParamObj();
	std::string side = par->getString("side", "right");
	int iShot = par->getInt("iShot");
    int nbPanel = par->getInt("nbPanel",3);

    // Observed data
    std::shared_ptr<SEP::genericRegFile> dObsFile = io->getRegFile("obsIn",usageIn);
	std::shared_ptr<SEP::hypercube> dObsHyper = dObsFile->getHyper();
	std::shared_ptr<SEP::float2DReg> dObs(new SEP::float2DReg(dObsHyper));
	dObsFile->readFloatStream(dObs);

    // Data residuals or predicted data
	int res = par->getInt("res",1);
    std::shared_ptr<SEP::genericRegFile> modelFile = io->getRegFile("model",usageIn);
	std::shared_ptr<SEP::hypercube> modelHyper = modelFile->getHyper();
	std::shared_ptr<SEP::float3DReg> model(new SEP::float3DReg(modelHyper));
	modelFile->readFloatStream(model);

    // Iteration
    int nIter=model->getHyper()->getAxis(3).n;
	axis iterAxis = axis(nIter, 0.0, 1.0);

    // Shots
    int oShotGrid=par->getInt("xSource");
	int dShotGrid=par->getInt("spacingShots");

    // Receivers
    float oRec=model->getHyper()->getAxis(2).o;
    float dRec=model->getHyper()->getAxis(2).d;
    int nRec=model->getHyper()->getAxis(2).n;
	axis recAxis = axis(nRec, oRec, dRec);

    // Time
    float ots=model->getHyper()->getAxis(1).o;
    float dts=model->getHyper()->getAxis(1).d;
    int nts=model->getHyper()->getAxis(1).n;
	axis timeAxis = axis(nts, ots, dts);

	// Find indices on the "shot grid" and "receiver grid"
	int iShotRecGrid=oShotGrid+iShot*dShotGrid;
	float xShotNew=oRec+iShotRecGrid*dRec;



    // Find the number of traces for each side
    int nRecNew;
	if(side=="right"){
		nRecNew=nRec-iShotRecGrid-1;
	} else{
		nRecNew=iShotRecGrid;
    }

    // Total number of receivers for the super shot gather
    if (nbPanel == 3){

        // Compute total number of receiver in super shot gather
        int nRecTotal = 3*nRecNew+2;

        // Allocate super shot gather
        axis recNewAxis=axis(nRecTotal,-nRecNew*dRec,dRec);
    	std::shared_ptr<hypercube> dataHyper(new hypercube(timeAxis, recNewAxis, iterAxis));
    	std::shared_ptr<SEP::float3DReg> data(new SEP::float3DReg(dataHyper));
        data->scale(0.0);

        // Fill data
        if (side == "right"){

            #pragma omp parallel for
            for (int iIter=0; iIter<nIter; iIter++){
                for (int iRec=0; iRec<nRecNew; iRec++){
                    for (int its=0; its<nts; its++){
                        (*data->_mat)[iIter][iRec][its] = (*dObs->_mat)[iShotRecGrid+nRecNew-iRec][its];
                        (*data->_mat)[iIter][iRec+nRecNew+1][its] = (*model->_mat)[iIter][iShotRecGrid+iRec+1][its]+(*dObs->_mat)[iShotRecGrid+iRec+1][its];
                        (*data->_mat)[iIter][nRecNew][its] = (*dObs->_mat)[iShotRecGrid][its];
                        (*data->_mat)[iIter][iRec+2*nRecNew+1][its] = (*dObs->_mat)[iShotRecGrid+nRecNew-iRec][its];
                    }
                }
            }

        } else {

            #pragma omp parallel for
            for (int iIter=0; iIter<nIter; iIter++){
                for (int iRec=0; iRec<nRecNew; iRec++){
                    for (int its=0; its<nts; its++){
                        (*data->_mat)[iIter][iRec][its] = (*dObs->_mat)[iRec][its];
                        (*data->_mat)[iIter][iRec+nRecNew+1][its] = (*model->_mat)[iIter][iShotRecGrid-iRec-1][its]+(*dObs->_mat)[iShotRecGrid-iRec-1][its];
                        (*data->_mat)[iIter][nRecNew][its] = (*dObs->_mat)[iShotRecGrid][its];
                        (*data->_mat)[iIter][iRec+2*nRecNew+1][its] = (*dObs->_mat)[iRec][its];
                    }
                }
            }
        }

        /* Files shits */
        std::shared_ptr <genericRegFile> dataFile;
        dataFile = io->getRegFile(std::string("data"), usageOut);
        dataFile->setHyper(dataHyper);
        dataFile->writeDescription();
    	dataFile->writeFloatStream(data);

    } else {

        // Compute total number of receiver in super shot gather
        int nRecTotal=2*nRecNew+1;

        // Allocate super shot gather
        axis recNewAxis=axis(nRecTotal,-nRecNew*dRec,dRec);
    	std::shared_ptr<hypercube> dataHyper(new hypercube(timeAxis, recNewAxis, iterAxis));
    	std::shared_ptr<SEP::float3DReg> data(new SEP::float3DReg(dataHyper));
        data->scale(0.0);

        // Fill data
        if (side == "right"){
            #pragma omp parallel for
            for (int iIter=0; iIter<nIter; iIter++){
                for (int iRec=0; iRec<nRecNew; iRec++){
                    for (int its=0; its<nts; its++){
                        (*data->_mat)[iIter][iRec][its] = (*dObs->_mat)[iShotRecGrid+nRecNew-iRec][its];
                        (*data->_mat)[iIter][iRec+nRecNew+1][its] = (*model->_mat)[iIter][iShotRecGrid+iRec+1][its]+(*dObs->_mat)[iShotRecGrid+iRec+1][its];
                        (*data->_mat)[iIter][nRecNew][its] = (*dObs->_mat)[iShotRecGrid][its];
                    }
                }
            }
        } else {
            #pragma omp parallel for
            for (int iIter=0; iIter<nIter; iIter++){
                for (int iRec=0; iRec<nRecNew; iRec++){
                    for (int its=0; its<nts; its++){
                        (*data->_mat)[iIter][iRec][its] = (*dObs->_mat)[iRec][its];
                        (*data->_mat)[iIter][iRec+nRecNew+1][its] = (*model->_mat)[iIter][iShotRecGrid-iRec-1][its]+(*dObs->_mat)[iShotRecGrid-iRec-1][its];
                        (*data->_mat)[iIter][nRecNew][its] = (*dObs->_mat)[iShotRecGrid][its];
                    }
                }
            }
        }

        /* Files shits */
        std::shared_ptr <genericRegFile> dataFile;
        dataFile = io->getRegFile(std::string("data"), usageOut);
        dataFile->setHyper(dataHyper);
        dataFile->writeDescription();
    	dataFile->writeFloatStream(data);

    }

    return 0;
}

File: acoustic_iso_lib/seis_utils/seis_utils_float/python/dataCompMain.py
#!/usr/bin/env python3
import genericIO
import SepVector
import Hypercube
import numpy as np
import sys

if __name__ == '__main__':

	# IO object
	parObject=genericIO.io(params=sys.argv)
	
	offset=parObject.getString("offset","pos")
	# xShot=parObject.getFloat("xShot") # Shot position [km]
	iShot=parObject.getInt("iShot")
	res=parObject.getInt("res",1) # Flag to determined whether the input model is a predicted data or data residual

	# Read true data
	obsDataFile=parObject.getString("obsIn")
	obsData=genericIO.defaultIO.getVector(obsDataFile,ndims=2)
	obsDataNp=obsData.getNdArray()

	# Read data residuals or predicted data
	modelFile=parObject.getString("model")
	model=genericIO.defaultIO.getVector(modelFile,ndims=3)
	modelNp=model.getNdArray()

	# Iteration
	oIter=model.getHyper().axes[2].o
	dIter=model.getHyper().axes[2].d
	nIter=model.getHyper().axes[2].n
	iterAxis=Hypercube.axis(n=nIter,o=oIter,d=dIter,label="Iteration #")

	# Shot
	oShotGrid=parObject.getInt("xSource")
	dShotGrid=parObject.getInt("spacingShots")

	# Receiver
	oRec=obsData.getHyper().axes[1].o
	dRec=obsData.getHyper().axes[1].d
	nRec=obsData.getHyper().axes[1].n
	recAxis=Hypercube.axis(n=nRec,o=oRec,d=dRec,label="Receivers [km]")

	# Time
	ots=obsData.getHyper().axes[0].o
	dts=obsData.getHyper().axes[0].d
	nts=obsData.getHyper().axes[0].n
	timeAxis=Hypercube.axis(n=nts,o=ots,d=dts,label="Time [s]")

	# Find indices on the "receiver grid"
	iShotRecGrid=oShotGrid+iShot*dShotGrid
	xShot=oRec+iShotRecGrid*dRec
	print("xShot=",xShot)
	print("iShot=",iShot)
	print("iShotRecGrid=",iShotRecGrid)

	# If user provides residual data instead of predicted data
	if(res==1):
		for iIter in range(nIter):
			modelNp[iIter][:][:]=modelNp[iIter][:][:]+obsDataNp[:][:]

	# Find the number of traces for each side
	if(offset=="pos"):
		nRecNew=nRec-iShotRecGrid-1
	else:
		nRecNew=iShotRecGrid

	# Total number of receivers for the super shot gather
	nRecTotal=2*nRecNew+1

	# Allocate super shot gather
	recNewAxis=Hypercube.axis(n=nRecTotal,o=-nRecNew*dRec,d=dRec,label="Offset [km]")
	iterAxis=Hypercube.axis(n=nIter,o=oIter,d=dIter,label="Iteration #")
	# shotNewAxis=Hypercube.axis(n=1,o=xShotNew,d=1.0)
	superShotGatherHyper=Hypercube.hypercube(axes=[timeAxis,recNewAxis,iterAxis])
	superShotGather=SepVector.getSepVector(superShotGatherHyper)
	superShotGather.scale(0.0)
	superShotGatherNp=superShotGather.getNdArray()

	# Allocate observed data
	# obsHyper=Hypercube.hypercube(axes=[timeAxis,recAxis])
	# obs=SepVector.getSepVector(obsHyper)
	# obsNp=obs.getNdArray()
	# predHyper=Hypercube.hypercube(axes=[timeAxis,recAxis,iterAxis])
	# pred=SepVector.getSepVector(predHyper)
	# predNp=pred.getNdArray()

	# Copy value to super shot gather
	if (offset=="pos"):
		for iIter in range(nIter):
			for iRec in range(nRecNew):
				for its in range(nts):
					superShotGatherNp[iIter][iRec][its]=obsDataNp[iShotRecGrid+nRecNew-iRec][its]
					superShotGatherNp[iIter][iRec+nRecNew+1][its]=modelNp[iIter][iShotRecGrid+iRec+1][its]
			superShotGatherNp[iIter][nRecNew][:]=obsDataNp[iShotRecGrid][:]
			# predNp[iIter][:][:]=modelNp[iIter][:][:]

	else:
		for iIter in range(nIter):
			for iRec in range(nRecNew):
				superShotGatherNp[iIter][iRec][:]=obsDataNp[iRec][:]
				superShotGatherNp[iIter][iRec+nRecNew+1][:]=modelNp[iIter][iShotRecGrid-iRec-1][:]
			superShotGatherNp[iIter][nRecNew][:]=obsDataNp[iShotRecGrid][:]
			# predNp[iIter][:][:]=modelNp[iIter][:][:]

	# obsNp[:][:]=obsDataNp[iShot][:][:]

	# Write super shot gather
	superShotGatherFile=parObject.getString("data")
	genericIO.defaultIO.writeVector(superShotGatherFile,superShotGather)

	# Write observed shot gather
	# obsOutFile=parObject.getString("obsOut")
	# genericIO.defaultIO.writeVector(obsOutFile,obs)
	#
	# # Write predicted shot gather
	# predFile=parObject.getString("pred")
	# genericIO.defaultIO.writeVector(predFile,pred)
